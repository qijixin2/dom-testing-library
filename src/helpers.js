const globalObj = typeof window === 'undefined' ? global : window
// Constant node.nodeType for text nodes, see:
// https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants
const TEXT_NODE = 3

// Currently this fn only supports jest timers, but it could support other test runners in the future.
function runWithRealTimers(callback) {
  return _runWithRealTimers(callback).callbackReturnValue
}

function _runWithRealTimers(callback) {
  const timerAPI = {
    clearImmediate,
    clearInterval,
    clearTimeout,
    setImmediate,
    setInterval,
    setTimeout,
  }

  // istanbul ignore else
  if (typeof jest !== 'undefined') {
    jest.useRealTimers()
  }

  const callbackReturnValue = callback()

  const usedJestFakeTimers = Object.entries(timerAPI).some(
    ([name, func]) => func !== globalObj[name],
  )

  if (usedJestFakeTimers) {
    jest.useFakeTimers(timerAPI.setTimeout?.clock ? 'modern' : 'legacy')
  }

  return {
    callbackReturnValue,
    usedJestFakeTimers,
  }
}

const jestFakeTimersAreEnabled = () =>
  Boolean(_runWithRealTimers(() => {}).usedJestFakeTimers)

// we only run our tests in node, and setImmediate is supported in node.
// istanbul ignore next
function setImmediatePolyfill(fn) {
  return globalObj.setTimeout(fn, 0)
}

function getTimeFunctions() {
  // istanbul ignore next
  return {
    clearTimeoutFn: globalObj.clearTimeout,
    setImmediateFn: globalObj.setImmediate || setImmediatePolyfill,
    setTimeoutFn: globalObj.setTimeout,
  }
}

const {clearTimeoutFn, setImmediateFn, setTimeoutFn} = runWithRealTimers(
  getTimeFunctions,
)

function getDocument() {
  /* istanbul ignore if */
  if (typeof window === 'undefined') {
    throw new Error('Could not find default container')
  }
  return window.document
}
function getWindowFromNode(node) {
  if (node.defaultView) {
    // node is document
    return node.defaultView
  } else if (node.ownerDocument && node.ownerDocument.defaultView) {
    // node is a DOM node
    return node.ownerDocument.defaultView
  } else if (node.window) {
    // node is window
    return node.window
  } else if (node.then instanceof Function) {
    throw new Error(
      `It looks like you passed a Promise object instead of a DOM node. Did you do something like \`fireEvent.click(screen.findBy...\` when you meant to use a \`getBy\` query \`fireEvent.click(screen.getBy...\`, or await the findBy query \`fireEvent.click(await screen.findBy...\`?`,
    )
  } else {
    // The user passed something unusual to a calling function
    throw new Error(
      `Unable to find the "window" object for the given node. Please file an issue with the code that's causing you to see this error: https://github.com/testing-library/dom-testing-library/issues/new`,
    )
  }
}

function checkContainerType(container) {
  if (
    !container ||
    !(typeof container.querySelector === 'function') ||
    !(typeof container.querySelectorAll === 'function')
  ) {
    throw new TypeError(
      `Expected container to be an Element, a Document or a DocumentFragment but got ${getTypeName(
        container,
      )}.`,
    )
  }

  function getTypeName(object) {
    if (typeof object === 'object') {
      return object === null ? 'null' : object.constructor.name
    }
    return typeof object
  }
}

export {
  getWindowFromNode,
  getDocument,
  clearTimeoutFn as clearTimeout,
  setImmediateFn as setImmediate,
  setTimeoutFn as setTimeout,
  runWithRealTimers,
  checkContainerType,
  jestFakeTimersAreEnabled,
  TEXT_NODE,
}
