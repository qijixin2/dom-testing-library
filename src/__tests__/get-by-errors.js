import cases from 'jest-in-case'
import {render} from './helpers/test-utils'

cases(
  'getBy* queries throw an error when there are multiple elements returned',
  ({name, query, html}) => {
    const utils = render(html)
    expect(() => utils[name](query)).toThrow(/multiple elements/i)
  },
  {
    getByLabelText: {
      query: /his/,
      html: `<div aria-label="his"></div><div aria-label="history"></div>`,
    },
    getByPlaceholderText: {
      query: /his/,
      html: `<input placeholder="his" /><input placeholder="history" />`,
    },
    getByText: {
      query: /his/,
      html: `<div>his</div><div>history</div>`,
    },
    getByAltText: {
      query: /his/,
      html: `<img alt="his" src="his.png" /><img alt="history" src="history.png" />`,
    },
    getByTitle: {
      query: /his/,
      html: `<div title="his"></div><div title="history"></div>`,
    },
    getByRole: {
      query: /his/,
      html: `<div role="his"></div><div role="history"></div>`,
    },
    getByTestId: {
      query: /his/,
      html: `<div data-testid="his"></div><div data-testid="history"></div>`,
    },
  },
)

cases(
  'queryBy* queries throw an error when there are multiple elements returned',
  ({name, query, html}) => {
    const utils = render(html)
    expect(() => utils[name](query)).toThrow(/multiple elements/i)
  },
  {
    queryByLabelText: {
      query: /his/,
      html: `<div aria-label="his"></div><div aria-label="history"></div>`,
    },
    queryByPlaceholderText: {
      query: /his/,
      html: `<input placeholder="his" /><input placeholder="history" />`,
    },
    queryByText: {
      query: /his/,
      html: `<div>his</div><div>history</div>`,
    },
    queryByAltText: {
      query: /his/,
      html: `<img alt="his" src="his.png" /><img alt="history" src="history.png" />`,
    },
    queryByTitle: {
      query: /his/,
      html: `<div title="his"></div><div title="history"></div>`,
    },
    queryByRole: {
      query: /his/,
      html: `<div role="his"></div><div role="history"></div>`,
    },
    queryByTestId: {
      query: /his/,
      html: `<div data-testid="his"></div><div data-testid="history"></div>`,
    },
  },
)

describe('*ByDisplayValue queries throw an error when there are multiple elements returned', () => {
  test('getByDisplayValue', () => {
    const {getByDisplayValue} = render(
      `<input value="his" /><select><option value="history">history</option></select>`,
    )
    expect(() => getByDisplayValue(/his/)).toThrow(
      /multiple elements with the display value:/i,
    )
  })
  test('queryByDisplayValue', () => {
    const {queryByDisplayValue} = render(
      `<input value="his" /><select><option value="history">history</option></select>`,
    )
    expect(() => queryByDisplayValue(/his/)).toThrow(
      /multiple elements with the display value:/i,
    )
  })
})
