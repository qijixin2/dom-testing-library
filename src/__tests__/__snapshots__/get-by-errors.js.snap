// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getByLabelText query will throw the custom error returned by config.getElementError 1`] = `"My custom error: Unable to find a label with the text of: TEST QUERY"`;

exports[`getByText query will throw the custom error returned by config.getElementError 1`] = `"My custom error: Unable to find an element with the text: TEST QUERY. This could be because the text is broken up by multiple elements. In this case, you can provide a function for your text matcher to make your matcher more flexible."`;
